### PR TITLE
Elixir support

### PR DESCRIPTION
**Pull request by nifoc**
This PR adds support for Elixir, thanks to [this awesome parser](https://github.com/ananthakumaran/tree-sitter-elixir).

I have been dogfooding this for quite some time (via the custom parser configuration) and didn't run into any issues.

A couple of important points:

* Makes use of the new `#any-of?` predicate
* I'm not sure if I have actually added `elixir` to alle the right places (mainly just `README.md` and `parsers.lua`)
* Will the entry in `lockfile.json` be generated automatically?
* I've been installing the parser via `grammer.js`. I'm not sure if I have to add this to the configuration or if having `elixir` shipped in `nvim-treesitter` will automatically pre-compile stuff?
* Since I intend to maintain the queries, I have added myself as the maintainer 🙂

### CODE REVIEW

File: README.md
Comment (theHamsta) [inline README.md:148]: Readme and lock file are autogenerated

File: lua/nvim-treesitter/parsers.lua
Comment (theHamsta) [inline lua/nvim-treesitter/parsers.lua:238]: You have to add a requires _generate_from_source. If it's not in the repo it needs to be generated. You also have to check whether npm is required (e.g. for depdencies or whether tree-sitter-cli is sufficient.

File: queries/elixir/highlights.scm
Comment (theHamsta) [inline queries/elixir/highlights.scm:182]: A very long expression

**Review by theHamsta — CHANGES_REQUESTED** [review]
Looks very nice. Probably  the config needs to be adapted so that the installation works automatically

File: queries/elixir/locals.scm
Comment (theHamsta) [inline queries/elixir/locals.scm:42]: Do you really need all this context

File: queries/elixir/highlights.scm
Comment (nifoc) [reply queries/elixir/highlights.scm:182]: Would it be better to split this up into multiple expressions (i.e. one expression per "case")?

Since you can do pattern matching in function heads in Elixir, function definitions can get quite complicated …

**Review by theHamsta — APPROVED** [review]

File: queries/elixir/highlights.scm
Comment (theHamsta) [reply queries/elixir/highlights.scm:182]: You are the maintainer. When you think this is the best way to work with it I support you :+1:

**Review by theHamsta — APPROVED** [review]

**Comment by theHamsta** [comment]
We had the discussion whether we should add filetype detect commands for languages not supported natively by nvim. What's your opinion on adding something like this?
https://github.com/theHamsta/nvim-treesitter/blob/efbb1c66d27eb5b4bfbcc1f59d3384e0641c8214/ftdetect/gdscript.vim#L1

Should this be in plugins/personal configs or nvim-treesitter?

**Comment by nifoc** [comment]
Just a quick thought: Maybe we should hold off merging this for a few more days, because the `any-of?` predicate is so new? (I saw you mentioned this in another PR)

Give people a bit more to install a newer nightly.

**Comment by theHamsta** [comment]
I also thought this #1295

**Comment by nifoc** [comment]
> We had the discussion whether we should add filetype detect commands for languages not supported natively by nvim. What's your opinion on adding something like this?
> https://github.com/theHamsta/nvim-treesitter/blob/efbb1c66d27eb5b4bfbcc1f59d3384e0641c8214/ftdetect/gdscript.vim#L1
> 
> Should this be in plugins/personal configs or nvim-treesitter?

Oh, right. I didn't think about that. When I uninstalled `vim-elixir` I added a similar line to my personal config.
But I can see why adding adding ftdetect for unsupported languages to nvim-treesitter might be beneficial.

Right now I don't have a strong opinion either way … But I will definitely give some thought to this!

File: queries/elixir/highlights.scm
Comment (vigoux) [reply queries/elixir/highlights.scm:182]: I am not sure it is better to split it in mutiple parts?

Comment (nifoc) [reply queries/elixir/highlights.scm:182]: I was playing around with this today and I feel like multiple parts are way more readable 👍 

I will update the PR accordingly.

**Comment by nifoc** [comment]
I'm sorry for the large amount of changes in my latest push. There was an update to the parser that simplified some things (regarding unused and _special_ variables) and I took the time to improve other queries a bit more.

I also added comments above some of the more complex queries, hopefully making it easier for other people to contribute.

**Comment by theHamsta** [comment]
At the moment your queries will not be tested. If you want you can install the tree-sitter CLI on CI to install also this parser.

File: scripts/ci-install-ubuntu-latest.sh
Comment (theHamsta) [inline scripts/ci-install-ubuntu-latest.sh:5]: :+1: 

Very strange though...
` Warning: parser for elixir is not installed`

File: scripts/ci-install-macos-latest.sh
Comment (theHamsta) [inline scripts/ci-install-macos-latest.sh:4]: This seems to be included already in `brew install neovim --HEAD`

Comment (nifoc) [reply scripts/ci-install-macos-latest.sh:4]: Have removed it locally

Comment (theHamsta) [reply scripts/ci-install-macos-latest.sh:4]: https://github.com/theHamsta/nvim-treesitter/blob/62bda244c9bf7e88f31d119f72bceae0e7116443/lua/nvim-treesitter/install.lua#L157

It does not work due to this line. Sorry for putting it there in the first place. I guess the logic should be if `not requires_generate_from_source or (not vim.fn.has('win32') and not generate_requires_npm)`

File: scripts/ci-install-ubuntu-latest.sh
Comment (nifoc) [reply scripts/ci-install-ubuntu-latest.sh:5]: It looks like the script thinks `npm` is required:

> Skipping language elixir on CI (requires npm)!

I haven't looked into why that is yet. I will take a look tomorrow.

File: scripts/ci-install-macos-latest.sh
Comment (theHamsta) [reply scripts/ci-install-macos-latest.sh:4]: @nifoc maybe I was wrong...   maybe just `requires_generate_from_grammar` -> `generate_requires_npm` as the message says.

Comment (theHamsta) [reply scripts/ci-install-macos-latest.sh:4]: I hate CIs... :sweat_smile:

Comment (nifoc) [reply scripts/ci-install-macos-latest.sh:4]: hmmm … If it change the line to `repo.generate_requires_npm` I think it would not install `typescript` (+ others)?
I only checked for `typescript` and it's currently installed in CI.

Comment (theHamsta) [reply scripts/ci-install-macos-latest.sh:4]: You're right... Can this check be removed completely?

File: scripts/ci-install-ubuntu-latest.sh
Comment (theHamsta) [reply scripts/ci-install-ubuntu-latest.sh:5]: That's the error message right after this check. But it does not fit very well to the check.

File: scripts/ci-install-macos-latest.sh
Comment (theHamsta) [reply scripts/ci-install-macos-latest.sh]: @nifoc this seems to have worked. All parsers are now either generatable without npm or installable directly from C files.

**Comment by theHamsta** [comment]
Nice work!

**Comment by dkarter** [comment]
Thanks for adding Elixir support @nifoc !

For anyone coming here because your Elixir syntax highlighting is suddenly broken with the "any-of" message, try upgrading Neovim 0.5 to the latest head version.

https://github.com/nvim-treesitter/nvim-treesitter/issues/1301

**Comment by duduvz14** [comment]
I noticed considerable lag when opening elixir files, and narrowed it down to when the highlight module is enabled, this is not happening with other files being parsed by treesitter, do you think this is something in my particular config? How could I help? Thanks for adding elixir support by the way!

**Comment by theHamsta** [comment]
@duduvz14 We observed some lags in some languages since 0.19.0 for parsing the queries in some languages https://github.com/tree-sitter/tree-sitter/issues/973 . Maybe you can try to reproduce that without Neovim. Would be great to pinpoint where the problem is.

**Pull request merged**
